pkg/ingester/ingester.go

```diff
@@ -3546,6 +3547,90 @@ func (i *Ingester) unsetPrepareShutdown() {
 	i.metrics.shutdownMarker.Set(0)
 }
 
+type prepareUnregisterBody struct {
+	Unregister *bool `json:"unregister"`
+}
+
+// PrepareUnregisterHandler manipulates whether an ingester will unregister from the ring on its next termination.
+//
+// The following methods are supported:
+//
+//   - GET
+//     Returns the ingester's current unregister state.
+//
+//   - PUT
+//     Sets the ingester's unregister state.
+//
+//   - DELETE
+//     Resets the ingester's unregister state to the value passed via the RingConfig.UnregisterOnShutdown ring
+//     configuration option.
+//
+// All methods are idempotent.
+func (i *Ingester) PrepareUnregisterHandler(w http.ResponseWriter, r *http.Request) {
+	if i.State() != services.Running {
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+
+	requestBody, err := io.ReadAll(r.Body)
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed to read prepare unregister request body", "err", err, "body", string(requestBody))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer func() {
+		if err := r.Body.Close(); err != nil {
+			level.Error(i.logger).Log("msg", "failed to close prepare unregister request body", "err", err)
+		}
+	}()
+
+	switch r.Method {
+	case http.MethodPut:
+		if len(requestBody) == 0 {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		input := prepareUnregisterBody{}
+		if err := json.Unmarshal(requestBody, &input); err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		if input.Unregister == nil {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		i.lifecycler.SetUnregisterOnShutdown(*input.Unregister)
+	case http.MethodGet:
+		if len(requestBody) > 0 {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+	case http.MethodDelete:
+		if len(requestBody) > 0 {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		i.lifecycler.SetUnregisterOnShutdown(i.cfg.IngesterRing.UnregisterOnShutdown)
+	default:
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	shouldUnregister := i.lifecycler.ShouldUnregisterOnShutdown()
+	responseBody, err := json.Marshal(&prepareUnregisterBody{Unregister: &shouldUnregister})
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed to marshal prepare unregister response body", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	_, _ = w.Write(responseBody)
```

Comment:
Let's replace this with

```suggestion
	util.WriteJSONResponse(w, &prepareUnregisterBody{Unregister: &shouldUnregister})
```

Reply:
Cool! I wasn't aware of the `util.WriteJSONResponse` function.

Action taken: suggestion accepted.